src/wallet/walletdb.cpp

```diff
@@ -1087,6 +1087,99 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
+        !pwallet->IsWalletFlagSet(WALLET_FLAG_GLOBAL_HD_KEY)
+        ) {
+        std::map<CExtPubKey, std::pair<std::map<std::string, int>, uint64_t>> descs_keys;
+        std::map<std::string, int> tmpl = {{"pkh(", 0}, {"sh(wpkh(", 0}, {"wpkh(", 0}};
+
+        // Find root xpubs used in pkh(), sh(wpkh()), and wpkh() descriptors
+        for (const auto& spkm : pwallet->GetAllScriptPubKeyMans()) {
+            const DescriptorScriptPubKeyMan* desc_spkm = dynamic_cast<DescriptorScriptPubKeyMan*>(spkm);
+            assert(desc_spkm);
+            LOCK(desc_spkm->cs_desc_man);
+            WalletDescriptor w_desc = desc_spkm->GetWalletDescriptor();
+
+            if (!w_desc.descriptor->IsRange()) {
+                continue;
+            }
+
+            std::set<CPubKey> desc_pubkeys;
+            std::set<CExtPubKey> desc_xpubs;
+            w_desc.descriptor->GetPubKeys(desc_pubkeys, desc_xpubs);
+            if (desc_xpubs.size() != 1) {
+                continue;
+            }
+            const CExtPubKey& xpub = *desc_xpubs.begin();
+
+            if (descs_keys.count(xpub) == 0) {
+                descs_keys.emplace(xpub, std::make_pair(tmpl, 0));
+            }
+
+            const std::string desc = w_desc.descriptor->ToString();
+            if (desc.find("pkh(") == 0) {
+                descs_keys[xpub].first["pkh("]++;
+            } else if (desc.find("sh(wpkh(") == 0) {
+                descs_keys[xpub].first["sh(wpkh("]++;
+            } else if (desc.find("wpkh(") == 0) {
+                descs_keys[xpub].first["wpkh("]++;
+            } else {
+                continue;
+            }
+            if (w_desc.creation_time > descs_keys[xpub].second) {
+                descs_keys[xpub].second = w_desc.creation_time;
+            }
+        }
+
+        // Find candidate active xpubs
+        // These are the ones that are used in 2 pkh(), 2 sh(wpkh()), and 2 wpkh() descriptors
+        uint64_t best_time = 0;
+        CExtPubKey best_xpub;
+        std::map<CKeyID, CExtPubKey> xpubs;
+        for (const auto& [xpub, info] : descs_keys) {
+            const auto& [dtypes, desc_time] = info;
+            if (std::all_of(dtypes.begin(), dtypes.end(), [](const auto& p) { return p.second == 2; })) {
+                xpubs.emplace(xpub.pubkey.GetID(), xpub);
+                if (desc_time > best_time) {
+                    best_time = desc_time;
+                    best_xpub = xpub;
+                }
+            }
+        }
+        if (best_time > 0) {
+            pwallet->SetActiveHDKey(best_xpub);
+        }
+
+        // Add hd keys
+        for (const auto& [id_pair, key] : wss.m_descriptor_keys) {
+            const auto& [desc_id, keyid] = id_pair;
+
+            const auto& it = xpubs.find(keyid);
+            if (it == xpubs.end()) {
+                continue;
+            }
+
+            CExtKey extkey(it->second, key);
+            pwallet->AddHDKey(extkey);
```

Comment:
Probably I don't follow, but it seems we don't need to add all descriptor keys as HD keys to the wallet

Reply:
I meant, that we can write only the key for `best_xpub`. Why do we need to write all the candidates as well?